### ISSUE TITLE
Handle errors better

### ISSUE THREAD

**Issue by Lukasa** [Lukasa/hyper#17]
Right now if `hyper` hits an error it just throws exceptions. This leaves the connection in an undefined state, and is fundamentally unusable from then on. We should start handling this better. 'Better' means two things:
- [ ] Errors on a single stream should cause us to send `RST_STREAM` frames. Exceptions should still be thrown, but the connection should be OK.
- [ ] Errors that affect the whole connection should cause us to send `GOAWAY` with an appropriate error code, and then tear the connection down.

**Comment by fredthomsen** [comment]
Would it be appropriate/preferred to define each error in RFC7540 with a corresponding exception?

**Comment by Lukasa** [comment]
@fredthomsen Nope, I don't think so. [This model](https://github.com/python-hyper/hyper-h2/blob/master/h2/exceptions.py#L17-L22) is better, I think.

**Comment by fredthomsen** [comment]
Ok. I'll take a crack at this.  I'll start a pull request once I have something.

**Comment by fredthomsen** [comment]
https://github.com/Lukasa/hyper/pull/188
